wallet-new/docs/updates.md

```diff
@@ -0,0 +1,563 @@
+<div id="table-of-contents">
+<h2>Table of Contents</h2>
+<div id="text-table-of-contents">
+<ul>
+<li><a href="#sec-1">1. Introduction</a></li>
+<li><a href="#sec-2">2. Abstract</a></li>
+<li><a href="#sec-3">3. Motivation</a></li>
+<li><a href="#sec-4">4. Current design</a>
+<ul>
+<li><a href="#sec-4-1">4.1. V0 Swagger spec</a></li>
+</ul>
+</li>
+<li><a href="#sec-5">5. Specification</a>
+<ul>
+<li><a href="#sec-5-1">5.1. New update system</a></li>
+<li><a href="#sec-5-2">5.2. Determining the severity</a></li>
+<li><a href="#sec-5-3">5.3. User stories</a></li>
+<li><a href="#sec-5-4">5.4. Installers retention</a></li>
+<li><a href="#sec-5-5">5.5. Storage &amp; data races</a></li>
+<li><a href="#sec-5-6">5.6. V1 Swagger spec</a></li>
+</ul>
+</li>
+</ul>
+</div>
+</div>
+
+# Introduction<a id="sec-1" name="sec-1"></a>
+
+# Abstract<a id="sec-2" name="sec-2"></a>
+
+The purpose of this document is to review how updates are fetched and delivered to
+the end user in Daedalus.
+
+# Motivation<a id="sec-3" name="sec-3"></a>
+
+When a new update is released to mainnet, Daedalus is able to inform the users
+about such update and guide them in the upgrade process. However, the current
+system suffers from a number of limitations, both in terms of functionality and
+ergonomics:
+
+-   Updates are silently downloaded from the network, unbeknownst to the users.
+    Users do not have visibility on this process up until the very last moment,
+    when an update prompt is presented to them.
+
+-   Although uploads can be postponed, they will be applied anyway when Daedalus
+    restarts, betraying users expectations.
+
+-   Users cannot dynamically query for updates, or really postpone them,
+    meaning that they are not in control of the update process.
+
+# Current design<a id="sec-4" name="sec-4"></a>
+
+This summarises (roughly) how the current update process works.
+NOTE: This is based on my own understanding (adinapoli) so the mileage
+might vary.
+
+-   The update is voted and it enters the blockchain.
+
+-   Nodes broadcasts this update between each other.
+
+-   When an edgenode receives an update, it register
+    this information internally.
+
+-   The wallet access this information and stores the
+    update in a list of updates. The list sorts the updates by
+    oldest first, so adding a new update is a `O(n)` operation where
+    `n` is the number of updates (which is small anyway).
```

Comment:
Ah! Is there any blocker in moving away from the current scheme but instead monitor the blockchain as we would do normally, but dispatch update blocks to this `ucDownloadedUpdate` variable without waiting for the full download? Am I missing something in the way we push updates to the blockchain? I was under the impression that updates are just normal blocks with special metadata inside which make them identifiable as updates, but this could totally be my distorted way of the world, so please let me know 😜

Reply:
@adinapoli-iohk Yes, we can totally populate this variable as soon as an update proposal appears in block, without downloading it; this is just about how things work right now. (And I think we should rename this variable, because it will signal not about downloaded update, but about update available to download).